File: 13 лаба/13C.cpp
#include <fstream>
#include <vector>

using namespace std;

int main() {

    ifstream fin("prefix.in");
    ofstream fout("prefix.out");

    string text;

    fin >> text;

    vector<int> prefix(text.length(),0);
    int j = 0;
    int i = 1;

    for(int i = 1; i < text.length(); i++){
        j = prefix[i-1];
        while((j>0)&&(text[j]!= text[i])){
            j = prefix[j-1];
        }
        if(text[j]==text[i]){
            j++;
        }
        prefix[i]=j;
    }

    for (int i=0; i< prefix.size(); i++){
        fout << prefix[i] << " ";
    }

    return 0;
}

File: 13 лаба/13C.py

fin = open("prefix.in",  "r")
fout = open("prefix.out", "w")

pattern = fin.readline()

prefix = [0]*(len(pattern)+1)
j = 0
i = 1

while i < len(pattern):

	if pattern[j]== pattern[i]:
		prefix[i+1]=j+1
		i+=1
		j+=1
		
	else:
		if j>0:
			j=prefix[j]
			
		else:
			prefix[i+1]=0
			i+=1
			
#print(prefix)	

for i in range(1, len(prefix)): 
	fout.write(str(prefix[i]))
	if i!=len(prefix)-1:
		fout.write(" ")	

#input()
fin.close()
fout.close()